src/daft-functions/src/coalesce.rs

```diff
@@ -0,0 +1,297 @@
+use common_error::{DaftError, DaftResult};
+use daft_core::{
+    prelude::{BooleanArray, DaftLogical, Field, Schema},
+    series::{IntoSeries, Series},
+    utils::supertype::try_get_supertype,
+};
+use daft_dsl::{
+    functions::{ScalarFunction, ScalarUDF},
+    ExprRef,
+};
+use serde::{Deserialize, Serialize};
+
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct Coalesce {}
+
+#[typetag::serde]
+impl ScalarUDF for Coalesce {
+    fn as_any(&self) -> &dyn std::any::Any {
+        self
+    }
+
+    fn name(&self) -> &'static str {
+        "coalesce"
+    }
+
+    fn to_field(&self, inputs: &[ExprRef], schema: &Schema) -> DaftResult<Field> {
+        match inputs {
+            [] => Err(DaftError::SchemaMismatch(
+                "Expected at least 1 input args, got 0".to_string(),
+            )),
+            [input] => {
+                let input_field = input.to_field(schema)?;
+                Ok(input_field)
+            }
+            _ => {
+                let first_field = inputs[0].to_field(schema)?;
+                let mut output_dtype = first_field.dtype.clone();
+
+                for input in inputs {
+                    let lhs = input.to_field(schema)?.dtype;
+                    let rhs = &first_field.dtype;
+                    output_dtype = try_get_supertype(&lhs, rhs)?;
+
+                    if try_get_supertype(&lhs, rhs).is_err() {
+                        return Err(DaftError::SchemaMismatch(format!(
+                            "All input fields must have the same data type. Got {lhs} and {rhs}"
+                        )));
+                    }
+                }
+                Ok(Field::new(first_field.name, output_dtype))
+            }
+        }
+    }
+
+    fn evaluate(&self, inputs: &[Series]) -> DaftResult<Series> {
+        match inputs.len() {
+            0 => Err(DaftError::ComputeError("No inputs provided".to_string())),
+            1 => Ok(inputs[0].clone()),
+            _ => {
+                let name = inputs[0].name();
+                let dtype = inputs[0].data_type();
+                let len = inputs[0].len();
```

Comment:
Have the children been rewritten somewhere to include the casts to the computed supertype? I am working through some of this, and would like to understand how the children get coerced.